python/tank/platform/software_launcher.py

```diff
@@ -239,38 +276,156 @@ def minimum_supported_version(self):
         The minimum software version that is supported by the launcher.
         Returned as a string, for example "2015" or "2015.3.sp3".
         Returns ``None`` if no constraint has been set.
-        Also see related method :meth:`~is_version_supported`.
         """
         # returns none by default, subclassed by implementing classes
         return None
 
-    ##########################################################################################
-    # abstract methods
+    @property
+    def products(self):
+        """
+        A list of product names limiting executable discovery.
 
-    def scan_software(self, versions=None):
+        Example::
+
+            ["Houdini", "Houdini Core", "Houdini FX"]
         """
-        Performs a scan for software installations.
+        return self._products
 
-        :param list versions: List of strings representing versions
-                              to search for. If set to None or [], search
-                              for all versions. A version string is
-                              DCC-specific but could be something
-                              like "2017", "6.3v7" or "1.2.3.52"
-        :returns: List of :class:`SoftwareVersion` instances
+    @property
+    def versions(self):
         """
-        raise NotImplementedError
+        A list of versions limiting executable discovery.
+
+        Example::
+
+            ["15.5.324", "16.0.1.322"]
+        """
+        return self._versions
+
+    ##########################################################################################
+    # abstract methods
 
     def prepare_launch(self, exec_path, args, file_to_open=None):
         """
         Prepares the given software for launch
 
         :param str exec_path: Path to DCC executable to launch
         :param str args: Command line arguments as strings
-        :param str file_to_open: (optional) Full path name of a file to open on launch
+        :param str file_to_open: (optional) Full path name of a file to open on
+            launch
         :returns: :class:`LaunchInformation` instance
         """
         raise NotImplementedError
 
+    def _scan_software(self):
+        """
+        Abstract method that must be implemented by subclasses to return a list
+        of ``SoftwareVersion`` objects on the local filesystem.
+
+        Typical implementations will use functionality such as ``glob`` to
+        locate all versions and variations of executables on disk and then
+        creating ``SoftwareVersion`` objects for each executable. This method
+        is called by the ``get_supported_software()`` method, and each
+        ``SoftwareVersion`` object returned is checked against the launcher's
+        lists of supported version and product variations via the
+        ``_is_supported`` method.
+        """
+
+        raise NotImplementedError
+
+    def _scan_software_with_expression(self, match_template, regular_expression):
```

Comment:
one more thought: This method is not software specific. I think a name like `_glob_and_match` or something like that would be a more approporiate description.